kernel/src/scan/log_replay.rs

```diff
@@ -37,80 +37,118 @@ struct LogReplayScanner {
-    removes: Vec<Remove>,
-    selection_vector: Option<Vec<bool>>,
-    // whether or not we are visiting commit json (=true) or checkpoint (=false)
+/// A visitor that deduplicates a stream of add and remove actions into a stream of valid adds. Log
+/// replay visits actions newest-first, so once we've seen a file action for a given (path, dvId)
+/// pair, we should ignore all subsequent (older) actions for that same (path, dvId) pair. If the
+/// first action for a given file is a remove, then that file does not show up in the result at all.
+struct AddRemoveDedupVisitor<'seen> {
+    seen: &'seen mut HashSet<FileActionKey>,
+    selection_vector: Vec<bool>,
     is_log_batch: bool,
 }
 
-const ADD_FIELD_COUNT: usize = 15;
+impl<'seen> AddRemoveDedupVisitor<'seen> {
+    /// Checks if log replay already processed this logical file (in which case the current action
+    /// should be ignored). If not already seen, register it so we can recognize future duplicates.
+    fn already_seen(&mut self, key: FileActionKey) -> bool {
+        // Note: each (add.path + add.dv_unique_id()) pair has a
+        // unique Add + Remove pair in the log. For example:
+        // https://github.com/delta-io/delta/blob/master/spark/src/test/resources/delta/table-with-dv-large/_delta_log/00000000000000000001.json
 
-impl AddRemoveVisitor {
-    fn new(selection_vector: Option<Vec<bool>>, is_log_batch: bool) -> Self {
-        AddRemoveVisitor {
-            selection_vector,
-            is_log_batch,
-            ..Default::default()
+        if self.seen.contains(&key) {
+            debug!(
+                "Ignoring duplicate ({}, {:?}) in scan, is log {}",
+                key.path, key.dv_unique_id, self.is_log_batch
+            );
+            true
+        } else {
+            debug!(
+                "Including ({}, {:?}) in scan, is log {}",
+                key.path, key.dv_unique_id, self.is_log_batch
+            );
+            if self.is_log_batch {
+                // Remember file actions from this batch so we can ignore duplicates as we process
+                // batches from older commit and/or checkpoint files. We don't track checkpoint
+                // batches because they are already the oldest actions and never replace anything.
+                self.seen.insert(key);
+            }
+            false
         }
     }
+
+    /// True if this row contains an Add action that should survive log replay. Skip it if the row
+    /// is not an Add action, or the file has already been seen previously.
+    fn is_valid_add<'a>(&mut self, i: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<bool> {
+        // Add will have a path at index 0 if it is valid; otherwise, if it is a log batch, we
+        // may have a remove with a path at index 4.
+        let (path, getters, is_add) = if let Some(path) = getters[0].get_str(i, "add.path")? {
+            (path, &getters[1..4], true)
+        } else if !self.is_log_batch {
```

Comment:
I don't quite understand `is_log_batch`. This is a visitor for log actions, so presumably everything is a log batch?

Might be worth a doc comment or a rename.

Reply:
ah yea maybe this could use some more clarity in the code but ill just jump in to say: `is_log_batch` is supposed to mean `is_commit_batch` (that is, a batch of actions from commits vs checkpoints)

Reply:
aha makes sense, ty!

Reply:
Yeah this is really old terminology -- "log file" in the Delta spec means a nnn.json file -- what we normally call "commit file" today.